proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy

```diff
@@ -0,0 +1,727 @@
+
+lemma schedule_valid_cur_vcpu[wp]:
+  "\<lbrace>valid_cur_vcpu and valid_idle\<rbrace>
+   (schedule :: (unit, unit) s_monad)
+   \<lbrace>\<lambda>_. valid_cur_vcpu\<rbrace>"
+  unfolding schedule_def allActiveTCBs_def
+  by (wpsimp wp: alternative_wp select_wp)
+
+crunches cancel_all_ipc, blocked_cancel_ipc, unbind_maybe_notification, cancel_all_signals,
+         bind_notification, fast_finalise, deleted_irq_handler, post_cap_deletion, cap_delete_one,
+         reply_cancel_ipc, cancel_ipc, update_waiting_ntfn, send_signal, send_ipc, send_fault_ipc,
+         receive_ipc, handle_fault, handle_interrupt, handle_vm_fault, handle_hypervisor_fault,
+         send_signal, do_reply_transfer, unbind_notification, suspend, cap_swap, bind_notification,
+         restart, reschedule_required, possible_switch_to, thread_set_priority, reply_from_kernel
+  for arch_state[wp]: "\<lambda>s. P (arch_state s)"
+  and cur_thread[wp]: "\<lambda>s. P (cur_thread s)"
+  (wp: mapM_x_wp_inv thread_set.arch_state select_wp crunch_wps
+   simp: crunch_simps possible_switch_to_def reschedule_required_def)
+
+lemma do_unbind_notification_arch_tcb_at[wp]:
+  "do_unbind_notification ntfnptr ntfn tcbptr \<lbrace>arch_tcb_at P t\<rbrace>"
+  unfolding set_bound_notification_def set_simple_ko_def
+  apply (wpsimp wp: set_object_wp get_object_wp get_simple_ko_wp thread_get_wp)
+  by (fastforce simp: pred_tcb_at_def obj_at_def get_tcb_def)
+
+lemma unbind_notification_arch_tcb_at[wp]:
+  "unbind_notification tcb \<lbrace>arch_tcb_at P t\<rbrace>"
+  unfolding unbind_notification_def
+  by wpsimp
+
+lemma bind_notification_arch_tcb_at[wp]:
+  "bind_notification tcbptr ntfnptr \<lbrace>arch_tcb_at P t\<rbrace>"
+  unfolding bind_notification_def set_bound_notification_def set_simple_ko_def
+  apply (wpsimp wp: set_object_wp get_object_wp get_simple_ko_wp)
+  by (fastforce simp: pred_tcb_at_def obj_at_def get_tcb_def)
+
+lemma unbind_maybe_notification_arch_tcb_at[wp]:
+  "unbind_maybe_notification ntfnptr \<lbrace>arch_tcb_at P t\<rbrace>"
+  unfolding unbind_maybe_notification_def
+  by wpsimp
+
+crunches blocked_cancel_ipc, cap_delete_one, cancel_signal
+  for arch_tcb_at[wp]: "arch_tcb_at P t"
+  (wp: crunch_wps simp: crunch_simps)
+
+lemma reply_cancel_ipc_arch_tcb_at[wp]:
+  "reply_cancel_ipc ntfnptr \<lbrace>arch_tcb_at P t\<rbrace>"
+  unfolding reply_cancel_ipc_def thread_set_def
+  apply (wpsimp wp: set_object_wp select_wp)
+  by (clarsimp simp: pred_tcb_at_def obj_at_def get_tcb_def)
+
+crunches cancel_ipc, send_ipc, receive_ipc
+  for arch_tcb_at[wp]: "arch_tcb_at P t"
+  (wp: crunch_wps simp: crunch_simps)
+
+lemma send_fault_ipc_arch_tcb_at[wp]:
+  "send_fault_ipc tptr fault \<lbrace>arch_tcb_at P t\<rbrace>"
+  unfolding send_fault_ipc_def thread_set_def Let_def
+  apply (rule validE_valid)
+  apply (rule hoare_seq_ext_skipE, solves wpsimp)+
+  apply (wpsimp wp: set_object_wp)
+  apply (clarsimp simp: pred_tcb_at_def obj_at_def get_tcb_def)
+  done
```

Comment:
`wp` gets stuck at `lookup_cap` because we haven't provided any full `wp` rules for it, but we do have a rule for `arch_tcb_at` and all the implications are just accumulated information that is not actually necessary for that to be preserved. `hoare_drop_imps` will get rid of these implications automatically and `hoare_vcg_all_lift_R` will deal with the quantifiers that are introduced in the case distinction:
```suggestion
  by (wpsimp wp: set_object_wp hoare_drop_imps hoare_vcg_all_lift_R
             simp: pred_tcb_at_def obj_at_def get_tcb_def)
```